### PR TITLE
Manipulate module & component name and producers metadata

### PR DESCRIPTION
**Pull request by pchickey**
This PR adds tools to view and modify the name and producer metadata in a module or component:

Producers section spec: https://github.com/WebAssembly/tool-conventions/blob/main/ProducersSection.md

* `wasm-encoder` now has a `ProducersEncoder`
* `wasm-tools metadata show` shows the module/component name and producers metadata in a wasm binary. `--json` option available for machine readability.
* `wasm-tools metadata add` rewrites a wasm binary to include a module/component `--name`, or add a field to the producers section `--language`, `--processed-by`, or `--sdk`.
* new crate `wasm-metadata` provides the above functionality as reusable code
* `wit-component` adds a name and producer section onto the adapter, shim, and fixups modules

Example output (note, I updated preview2 prototyping to use this PR's version of wit-component so more metadata is present)
```
$ wasm-tools metadata show /home/phickey/src/preview2-prototyping/target/debug/build/test-programs-macros-7836162fb5a6267f/out/hello_stdout.component.wasm
component:
    processed-by:
        wit-component: 0.5.0
    module:
        language:
            Rust
        processed-by:
            clang: 15.0.6
            rustc: 1.69.0-nightly (11d96b593 2023-02-01)
    module wit-component:adapter:wasi_snapshot_preview1:
        language:
            Rust
        processed-by:
            rustc: 1.67.0 (fc594f156 2023-01-24)
    module wit-component:shim:
        processed-by:
            wit-component: 0.5.0
    module wit-component:fixups:
        processed-by:
            wit-component: 0.5.0
```

<details>
<summary>Json output</summary>
piped through jq for readability
<pre>
{
  "component": {
    "name": null,
    "producers": [
      [
        "processed-by",
        {
          "wit-component": "0.5.0"
        }
      ]
    ],
    "children": [
      {
        "module": {
          "name": null,
          "producers": [
            [
              "language",
              {
                "Rust": ""
              }
            ],
            [
              "processed-by",
              {
                "clang": "15.0.6",
                "rustc": "1.69.0-nightly (11d96b593 2023-02-01)"
              }
            ]
          ]
        }
      },
      {
        "module": {
          "name": "wit-component:adapter:wasi_snapshot_preview1",
          "producers": [
            [
              "language",
              {
                "Rust": ""
              }
            ],
            [
              "processed-by",
              {
                "rustc": "1.67.0 (fc594f156 2023-01-24)"
              }
            ]
          ]
        }
      },
      {
        "module": {
          "name": "wit-component:shim",
          "producers": [
            [
              "processed-by",
              {
                "wit-component": "0.5.0"
              }
            ]
          ]
        }
      },
      {
        "module": {
          "name": "wit-component:fixups",
          "producers": [
            [
              "processed-by",
              {
                "wit-component": "0.5.0"
              }
            ]
          ]
        }
      }
    ]
  }
}
</pre>
</details>

### CODE REVIEW

**Comment by alexcrichton** [comment]
> convenient way to modify a binary by adding fields to a producers section (only one section allowed per binary)

What about:

```
wasm-tools producers --add-language rust=1.0.0 foo.wasm > bar.wasm
wasm-tools producers --add-processed-by rustc=1.0.0 foo.wasm -o bar.wasm
wasm-tools producers --add-sdk foo=1.0.0 --add-language bar=2.0.0 foo.wasm | wasm-tools print
```

where with no `--add-*` arguments it prints the section, but with `--add-*` arguments it outputs a new binary? Sort of like `tar` where the same tool both extracts and creates.

> tests

Right now the CLI doesn't have any tests at all, but as the CLI grows over time the need for tests is sort of getting more pressing over time. I don't think it should necessarily be entirely up to you to create tests for a relatively minor feature like this. I think it's ok if you'd prefer to defer adding tests at a later date.

**Comment by pchickey** [comment]
Yes! I was thinking right along those lines.

As for tests, I mean that my encoder should have unit tests. I wrote it last thing friday evening and once I filled in all my todos without typechecking I called it a weekend, so this morning I'll try actually executing it :)

File: Cargo.toml [hunk 1]
```diff
@@ -147,3 +151,4 @@ strip = ['wasm-encoder', 'wasmparser', 'regex']
 compose = ['wasm-compose']
 demangle = ['rustc-demangle', 'cpp_demangle', 'wasmparser', 'wasm-encoder']
 component = ['wit-component', 'wit-parser', 'wast', 'wasm-encoder', 'wasmparser']
+producers = ['wasmparser', 'indexmap']
```
Comment (alexcrichton) [inline Cargo.toml:154]: Oh before I forget can you update `.github/workflows/main.yml` to test the build with just the `producers` feature? There should be a spot that all the other subcommands are tested as well.

**Review by alexcrichton — APPROVED** [review]

File: README.md [hunk 2]
```diff
@@ -55,6 +55,8 @@ programmatically as well:
 | `wasm-tools component new` | [wit-component] | Create a component from a core wasm binary |
 | `wasm-tools component wit` |  | Extract a `*.wit` interface from a component |
 | `wasm-tools component embed` |  | Embed a `component-type` custom section in a core wasm binary |
+| `wasm-tools metadata show` |  | Show name and producer metadata in a component or module |
```
Comment (alexcrichton) [inline README.md:58]: You can call out the wasm-metadata crate here between the `|  |`

**Review by alexcrichton — APPROVED** [review]
Looks great to me, I think this shaped up really well! 

Can you add some simple tests for `crates/wasm-metadata`? I'm hoping that given all the conveniences there it wouldn't be too bad and it could help ensure that recursive producers all pop out, adding a producers section to a component works, etc. I was a little worried reading the code that adding a producers section to a component with core wasm inside of it wouldn't work but tests would help assuage that.

File: README.md (hunk 2 above)
Comment (pchickey) [reply README.md]: Thanks

**Comment by pchickey** [comment]
Great idea! I added three tests - trivial module add metadata, trivial component add metadata, and then the case you were worried about.

There is no wat text format for this custom section, so unfortunately writing the tests is harder than it ought to be.

**Comment by pchickey** [comment]
We'll need to bump some package versions and publish a new wasm-encoder to publish wasm-metadata to crates.io.

**Comment by alexcrichton** [comment]
Doing that in https://github.com/bytecodealliance/wasm-tools/pull/907

**Pull request merged**